cleaner/scheduled_tasks/classes/clean.php

```diff
@@ -56,27 +59,29 @@ static public function execute() {
 
             // Disable every task that has a record in our table.
             foreach ($disabledtasks as $disabledtask) {
-                foreach ($tasknames as $taskname) {
-                    if ($taskname->id == $disabledtask->task_scheduled_id) {
-                        if ($taskname->disabled == 1) {
-                            mtrace("Task $increment/$count: $taskname->classname selected to disable but is already disabled, skipping..");
-                            $increment++;
-                        } else {
-                            if ($dryrun) {
-                                mtrace("Task $increment/$count: Would disable task: $taskname->classname");
-                                $increment++;
-                            } else {
-                                mtrace("Task $increment/$count: Disabling task: $taskname->classname");
-                                $updatetask = new \stdClass();
-                                $updatetask->id = $disabledtask->task_scheduled_id;
-                                $updatetask->disabled = 1;
-                                $DB->update_record('task_scheduled', $updatetask);
-                                $increment++;
-                            }
-                        }
+                if ($disabledtask->disabled == 1) {
+                    mtrace("Task $increment/$count: $disabledtask->classname selected to disable but is already disabled, skipping..");
+                    $increment++;
+                } else {
+                    if ($dryrun) {
+                        mtrace("Task $increment/$count: Would disable task: $disabledtask->classname");
+                        $increment++;
+                    } else {
+                        mtrace("Task $increment/$count: Disabling task: $disabledtask->classname");
+                        $updatetask = [];
+                        $updatetask['id'] = $disabledtask->taskscheduledid;
+                        $updatetask['disabled'] = 1;
+                        $taskstoupdate[] = $updatetask;
+                        //$DB->update_record('task_scheduled', $updatetask);
+                        $increment++;
                     }
                 }
             }
+            if ($taskstoupdate) {
+                // Collect all the tasks to update and update in one query
+                list($sql, $params) = $DB->get_in_or_equal($taskstoupdate);
+                $DB->execute("UPDATE {task_scheduled} SET disabled = 1 WHERE id " . $sql, $params);
```

Comment:
@dkleto This part wasn't working for me, which is why it's been reverted in the current version. The code expects id to be a number, where we pass it an array of numbers. Elsewhere in moodle I've seen this setup (get_in_or_equal followed by execute) be used with a foreach loop for each $param, which defeats the purpose of collecting them. There may be something I'm not getting here

Reply:
set_field_select might be what you're looking for here